### DISCUSSION TITLE
help adding and deleting nested array elements in v4

### DISCUSSION THREAD

**Discussion by olets** [mikefarah/yq#739] [Q&A] [answered]
I'm moving my project [git-replay](https://github.com/olets/git-replay) from yq v3 to v4, and am having trouble finding the correct new syntax for two things.

Working with a file like

```yaml
# myfile.yaml
array_1:
  array_2:
    - array_element_1
    - array_element_2
```

in v3 I would use

```shell
yq d myfile.yaml array_1.array_2.array_element_1
```

to delete `array_element_1` and 

```shell
yq w myfile.yaml "array_1.array_2[+]" array_element_3
```

to add `array_element_3`. 

What's the best v4 approach? Thanks!

**Comment by mikefarah** [comment, marked as answer]
https://mikefarah.gitbook.io/yq/operators/delete#delete-nested-entry-in-array

https://mikefarah.gitbook.io/yq/operators/add#add-string-to-array

**Reply by olets** [reply]
Thanks for your help and for this great tool!

For deleting I'm trying to delete a string array entry by value. For example delete `c` in

```
a:
  b:
    - c
    - d
```

I've tried `del(.a.b.c)`, `del([0].a.b.c)`, `del(.a.b.c.[0])`, `del([0].a.b.c.[0])`, and `del(.a.b["c"])` and all give a "Cannot index array with &lt;the problem node> (strconv.ParseInt: parsing &lt;the problem node>: invalid syntax)". I've also tried `del(.a.b.[0].c)` and that does not error but it returns the input unchanged. I've also tried `del(.a.b.[0].["c"])`, `del(.a.b["c"])`, and `.a.b = .a.b - "c"` all of which are errors.

Is it possible to look up an array item's index? `del(.a.b[i])` and `del(.a.b.[i])` where `i` is the string element's index might work if I could get `i` (I'm not currently requiring uniqueness but I could). I have not found that in the docs.

---

~For anyone else reading, the add solution to my problem combines the "add string to array" and "update array (append)" examples: `yq e '.a.b = .a.b + "e"' input`~

edit: My add solution will not work

```
#myfile
a:
  b:
    - c
```

`yq e '.a.d = .a.d + "e"' myfile --inplace`


```
#myfile
a:
  b:
    - c
  d: e
```

`yq e '.a.d = .a.d + "f"' myfile --inplace`


```
#myfile
a:
  b:
    - c
  d: ef
```

where the goal is 

```
#myfile
a:
  b:
    - c
  d:
    - e
    - f
```

`.a.d[0] = e` followed by `.a.d[1] = f` would work. Is there an array append? Tried `.a.d[+] =` and `.a.d +=`. Or is it possible to get the length of the array? Tried `length(.a.b)`, `length .a.b` and `.a.b length`.

**Reply by mikefarah** [reply]
I'm getting confused - I think this should have been two posts.

To delete 'c' in the a.b. array:

```
yq e 'del(.a.b.[] | select(.=="c"))' examples/data1.yaml
```

This basically splats the a.b array, then selects the matching elements to delete.

To do the append thing, the issue is it has no value to start with, so when you += a scalar (string/number) it has no way of knowing that what you want is an array.

You can fix this by either wrapping it in an array:

```
yq e '.a.d = .a.d + ["e"]' myfile --inplace
```

or providing a default value when it's not defined:

```
yq e '.a.d = (.a.d // []) + "e"'
```

**Reply by olets** [reply]
> I think this should have been two posts.

Fair. I'll move follow ups to new questions.

Thanks for the help!
